lib/src/front_end/ast_node_visitor.dart

```diff
@@ -1214,104 +1254,97 @@ class AstNodeVisitor extends ThrowingAstVisitor<void>
   }
 
   @override
-  void visitShowCombinator(ShowCombinator node) {
-    assert(false, 'Combinators are handled by createImport().');
+  Piece visitShowCombinator(ShowCombinator node) {
+    throw UnsupportedError('Combinators are handled by createImport().');
   }
 
   @override
-  void visitSimpleFormalParameter(SimpleFormalParameter node) {
-    startFormalParameter(node);
-
-    if ((node.type, node.name) case (var type?, var name?)) {
-      // Have both a type and name, so allow splitting between them.
-      modifier(node.keyword);
-      visit(type);
-      var typePiece = pieces.split();
+  Piece visitSimpleFormalParameter(SimpleFormalParameter node) {
+    var builder = AdjacentBuilder(this);
+    startFormalParameter(node, builder);
+    builder.modifier(node.keyword);
+    builder.visit(node.type);
 
-      token(name);
-      var namePiece = pieces.take();
-
-      pieces.give(VariablePiece(typePiece, [namePiece], hasType: true));
+    if ((node.type, node.name) case (var _?, var name?)) {
+      // Have both a type and name, so allow splitting after the type.
+      var typePiece = builder.build();
+      var namePiece = tokenPiece(name);
+      return VariablePiece(typePiece, [namePiece], hasType: true);
     } else {
-      // Only one of name or type so just write whichever there is.
-      modifier(node.keyword);
-      visit(node.type);
-      token(node.name);
+      // Don't have both a type and name, so just write whichever one we have.
```

Comment:
Does this write the type if that is all we have?
Should this read "just write the name"?

Reply:
> Does this write the type if that is all we have?

Yes, as in:

```dart
typedef TakeString = void Function(String);
```